core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,8 +62,14 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // visible for testing
+  private[spark] val isPartitionEvalParallel: Boolean =
+    rdds.length > conf.getInt("spark.rdd.parallelListingThreshold", 10)
```

Comment:
I don't agree that such a source would necessarily break other places in Spark. There's a big difference between threads may happen to execute some method concurrently and kicking off a pool of threads at that method.

While there is no known instance, this is based on two real-world cases:
1. Parquet had a caching scheme that assumed no reuse at all, which broken when that assumption was violated by Hive.
2. Parquet had a proposal to cache splits for Spark, which would require a static/class-level cache.

InputFormats are one of the most common places to plug in bad code because they are used for old, custom formats. I'm fine with parallelizing this by default, but I'm not comfortable with the idea of changing how they are used  and increasing the chances of hitting a bug if one exists without a safety valve. I also don't see the down-side to having one.